### ISSUE TITLE
max.in.flight should be enforced per partition, not connection

### ISSUE THREAD

**Issue by edenhill** [edenhill/librdkafka#1969]
max.in.flight.. provides a mean to provide guaranteed ordering without the idempotent producer by making sure only one outstanding ProduceRequest is in-flight to the broker at any given time.
But since ordering is only a thing on the partition-level in Kafka, not across partitions, there is no point to limit max.in.flight to per-connection, but it should be enforced per-partition instead.

This will increase performance and reduce latency for max.in.flight=1 producers producing to multiple partitions.

**Comment by saabeilin** [comment]
I'd say options shared with native Java client _should_ behave identically; how exactly is that implemented in Kafka Java client libraries?

**Comment by edenhill** [comment]
The Java client does it per connection, which the full name `max.in.flight.requests.per.connection` implies, the shorter `max.in.flight` was added to librdkafka, which means we could add a `max.in.flight.requests.per.partition` that does what we want, and possibly have the `max.in.flight` realiased to that.

**Comment by faladu** [comment]
I wanted to ask if something like the suggested max.in.flight.requests.per.partition exists?
And just to be sure, without setting that to 1 the order is not guaranteed when writing to a topic correct?

**Comment by edenhill** [comment]
Ordering is a partition-level concept in Kafka, not topic-level.

**Comment by faladu** [comment]
Yeah per partition is all I need.
So assuming I use rd_kafka_produce would I need to set max.in.flight.requests.per.connection to 1 to guarantee per partition ordering or would it work without it or some other setting?

In my current scenario i work with a one partition topic but the order received is completly different from the order sent. The only thing that seems to help is to wait for the callback before sending the next message but that kills the performance.

**Comment by edenhill** [comment]
Prior to v1.0.0 you would need to set max.in.flight=1, yes.
With v1.0.0 (soon to be released) you can enable the idempotent producer that will guarantee ordering, by setting `enable.idempotence=true` and handling ERR__FATAL errors.

**Comment by faladu** [comment]
Thank you for the information

Found my error i somehow had managed to have 2 topic handles for the same topic and due to that had a  strange ordering.

You say with v1.0.0 but the current librdkakfa source code already states version 1.0.8 so would this work already when i compile the binaries myself?

**Comment by edenhill** [comment]
1.0.8? Where do you see that?

master branch contains idempotent producer.

**Comment by faladu** [comment]
I'm sorry i mixed things up it states v1.0.0-RC8 not 1.0.8
I tried the idempotent producer as that is all i need however my broker were to old (got the error that the brokers don't support idempotent) so I'll be updating my kafka and then test some more.

Thank you a lot for all your responses.

**Comment by kenneth-jia** [comment]
The `idempotent producer` is quite a good feature, which becomes a MUST for some critical scenarios. With that feature enabled, customers' main concern would be the `performance`. The `max.in.flight=5` limitation for `per.connection` sounds not quite good, could highly impact the throughput.
I believe `max.in.flight.requests.per.partition` would be a much better choice, and consider the number of partitions within each connection, we could benefit from quite an impressive performance boost.
Really looking forward to that `max.in.flight.requests.per.partition`!

**Comment by KahnCheny** [comment]
> I'd say options shared with native Java client _should_ behave identically; how exactly is that implemented in Kafka Java client libraries?

I agree with you, because kafka broker processes sequentially on a single network connection.

**Comment by KahnCheny** [comment]
I propose that the default value of max.in.flight.requests.per.connection & max.in.flight  should be aligned with Java Client
